lib/wasi/src/runtime/module_cache/types.rs

```diff
@@ -56,14 +57,14 @@ pub trait ModuleCache: Debug {
     /// };
     ///
     /// let cache = SharedCache::default()
-    ///     .and_then(FileSystemCache::new("~/.local/cache"));
+    ///     .with_fallback(FileSystemCache::new("~/.local/cache"));
     /// ```
-    fn and_then<C>(self, other: C) -> AndThen<Self, C>
+    fn with_fallback<C>(self, other: C) -> FallbackCache<Self, C>
```

Comment:
I still think this should not be on the trait.

You recently argued with `Iterator` - this doesn't hold IMO.
Constructing a cache is a rare operation, not something you do constnatly.

This should just be `FallbackCache::new()`.

Reply:
I disagree. 

If you search for any uses of `SharedCache::default()` or `module_cache::in_memory()` you'll find they're almost all followed by a `with_fallback()`. Having useful combinator methods like this doesn't cost us anything and makes it a lot easier to implement layered caching.

Compare this:

```rust
SharedCache::default()
  .with_fallback(FileSystemCache::new(...))
  .with_fallback(DistributedCache::new(...));
```

with this:

```rust
let shared = SharedCache::new();
let filesystem = FileSystemCache::new(...);
let distributed = DistributedCache::new(...);
let cache = FallbackCache::new(shared, FallbackCache::new(filesystem, distributed));
```

Using method syntax both makes it a lot cleaner to write and makes it immediately obvious what the fallback order is.